core/node/api_server/src/execution_sandbox/apply.rs

```diff
@@ -37,179 +40,250 @@ use zksync_types::{
+        resolved_block_info.state_l2_block_number,
+        false,
+    )
+    .await
+    .context("cannot create `PostgresStorage`")?
+    .with_caches(setup_args.caches.clone());
+
+    let (system, l1_batch) = prepare_env(setup_args, &resolved_block_info, next_block);
+
+    let env = OneshotEnv {
+        system,
+        l1_batch,
+        current_block,
+    };
+    initialization_stage.observe();
+    Ok((env, storage))
+}
+
+async fn load_l2_block_info(
+    connection: &mut Connection<'_, Core>,
+    is_pending_block: bool,
+    resolved_block_info: &ResolvedBlockInfo,
+) -> anyhow::Result<(L2BlockEnv, Option<StoredL2BlockEnv>)> {
+    let mut current_block = None;
+    let next_block = read_stored_l2_block(connection, resolved_block_info.state_l2_block_number)
         .await
         .context("failed reading L2 block info")?;
 
-        let next_l2_block_info = if is_pending_block {
-            L2BlockEnv {
-                number: current_l2_block_info.l2_block_number + 1,
-                timestamp: resolved_block_info.l1_batch_timestamp,
-                prev_block_hash: current_l2_block_info.l2_block_hash,
-                // For simplicity, we assume each L2 block create one virtual block.
-                // This may be wrong only during transition period.
-                max_virtual_blocks_to_create: 1,
-            }
-        } else if current_l2_block_info.l2_block_number == 0 {
-            // Special case:
-            // - For environments, where genesis block was created before virtual block upgrade it doesn't matter what we put here.
-            // - Otherwise, we need to put actual values here. We cannot create next L2 block with block_number=0 and `max_virtual_blocks_to_create=0`
-            //   because of SystemContext requirements. But, due to intrinsics of SystemContext, block.number still will be resolved to 0.
-            L2BlockEnv {
-                number: 1,
-                timestamp: 0,
-                prev_block_hash: L2BlockHasher::legacy_hash(L2BlockNumber(0)),
-                max_virtual_blocks_to_create: 1,
-            }
-        } else {
-            // We need to reset L2 block info in storage to process transaction in the current block context.
-            // Actual resetting will be done after `storage_view` is created.
-            let prev_l2_block_info = StoredL2BlockInfo::new(
-                connection,
-                resolved_block_info.state_l2_block_number - 1,
-                None,
-            )
+    let next_block = if is_pending_block {
+        L2BlockEnv {
+            number: next_block.number + 1,
+            timestamp: resolved_block_info.l1_batch_timestamp,
+            prev_block_hash: resolved_block_info.state_l2_block_hash,
+            // For simplicity, we assume each L2 block create one virtual block.
+            // This may be wrong only during transition period.
+            max_virtual_blocks_to_create: 1,
+        }
+    } else if next_block.number == 0 {
```

Comment:
IIUC, this check has a false positive if the use has actually requested the genesis block. This wasn't covered by tests previously, but now with mock oneshot executor being more accurate, it was uncovered by VM tests. I'm not sure how to fix this, so I've hacked the tests instead by creating an additional block 🙃

Reply:
I think it's fine for now.